src/common/dao/base.go

```diff
@@ -187,3 +187,27 @@ func Escape(str string) string {
 	str = strings.Replace(str, `_`, `\_`, -1)
 	return str
 }
+
+// WithTransaction helper for transaction
+func WithTransaction(fn func(o orm.Ormer) error) error {
+	o := orm.NewOrm()
+
+	if err := o.Begin(); err != nil {
+		log.Errorf("begin transaction failed: %v", err)
+		return err
+	}
+
+	if err := fn(o); err != nil {
+		if err := o.Rollback(); err != nil {
```

Comment:
We should declare a new variable for the error returned by `Rollback` or the `err` returned by `fn` will be overwritten by `Rollback`. The error of `fn` should be returned.

Reply:
The err returns by `Rollback` will not overwritten the err returns by `fn`, see more https://goplay.space/#RuSSkIvv26J

But I think that when `Rollback` failed, we should return the error of `Rollback` which is an internal error. Otherwise, return the error of `fn`.